internal/gatewayapi/contexts.go

```diff
@@ -527,40 +209,100 @@ type TCPRouteContext struct {
+	hostnames := make([]string, h.Len())
+	for i := 0; i < len(hostnames); i++ {
+		hostnames[i] = h.Index(i).String()
+	}
+	return hostnames
 }
 
-func (t *TCPRouteContext) GetParentReferences() []v1beta1.ParentReference {
-	parentReferences := make([]v1beta1.ParentReference, len(t.Spec.ParentRefs))
-	for idx, p := range t.Spec.ParentRefs {
-		parentReferences[idx] = UpgradeParentReference(p)
+// TODO: [v1alpha2-v1beta1] This should not be required once all Route
+// objects being implemented are of type v1beta1.
+// GetParentReferences returns the ParentReference of the Route object.
+func GetParentReferences(route RouteContext) []v1beta1.ParentReference {
+	rv := reflect.ValueOf(route)
+	rt := rv.Type().String()
+	pr := rv.Elem().FieldByName("Spec").FieldByName("ParentRefs")
+	if strings.Contains(rt, "HTTP") || strings.Contains(rt, "GRPC") {
+		return pr.Interface().([]v1beta1.ParentReference)
+	}
+
+	parentReferences := make([]v1beta1.ParentReference, pr.Len())
+	for i := 0; i < len(parentReferences); i++ {
+		p := pr.Index(i).Interface().(v1beta1.ParentReference)
+		parentReferences[i] = UpgradeParentReference(p)
 	}
 	return parentReferences
 }
 
-func (t *TCPRouteContext) GetRouteStatus() *v1beta1.RouteStatus {
-	return &t.Status.RouteStatus
+// GetRouteStatus returns the RouteStatus object associated with the Route.
+func GetRouteStatus(route RouteContext) *v1beta1.RouteStatus {
+	rv := reflect.ValueOf(route).Elem()
+	rs := rv.FieldByName("Status").FieldByName("RouteStatus").Interface().(v1beta1.RouteStatus)
+	return &rs
 }
 
-func (t *TCPRouteContext) GetRouteParentContext(forParentRef v1beta1.ParentReference) *RouteParentContext {
-	if t.parentRefs == nil {
-		t.parentRefs = make(map[v1beta1.ParentReference]*RouteParentContext)
+// GetRouteParentContext returns RouteParentContext by using the Route
+// objects' ParentReference.
+func GetRouteParentContext(route RouteContext, forParentRef v1beta1.ParentReference) *RouteParentContext {
+	rv := reflect.ValueOf(route).Elem()
+	pr := rv.FieldByName("ParentRefs")
+	if pr.IsNil() {
+		mm := reflect.MakeMap(reflect.TypeOf(map[v1beta1.ParentReference]*RouteParentContext{}))
+		pr.Set(mm)
 	}
 
-	if ctx := t.parentRefs[forParentRef]; ctx != nil {
+	if p := pr.MapIndex(reflect.ValueOf(forParentRef)); p.IsValid() && !p.IsZero() {
+		ctx := p.Interface().(*RouteParentContext)
 		return ctx
 	}
 
+	isHTTPRoute := false
+	if strings.Contains(rv.Type().String(), "HTTP") {
+		isHTTPRoute = true
+	}
+
 	var parentRef *v1beta1.ParentReference
-	for i, p := range t.Spec.ParentRefs {
-		p := UpgradeParentReference(p)
-		if reflect.DeepEqual(p, forParentRef) {
-			upgraded := UpgradeParentReference(t.Spec.ParentRefs[i])
-			parentRef = &upgraded
+	specParentRefs := rv.FieldByName("Spec").FieldByName("ParentRefs")
+	for i := 0; i < specParentRefs.Len(); i++ {
+		p := specParentRefs.Index(i).Interface().(v1beta1.ParentReference)
+		up := p
+		if !isHTTPRoute {
```

Comment:
curious why this custom `HTTPRoute` logic is needed

Reply:
becasue there are some differences between the implementation of `HTTPRoute` and other Routes in `GetRouteParentContext`

for example, `HTTPRoute` is like:

https://github.com/envoyproxy/gateway/blob/7f8a4913b78b492c192b85262344dce4f4f64314/internal/gatewayapi/contexts.go#L220-L226

and other Route is like:

https://github.com/envoyproxy/gateway/blob/7f8a4913b78b492c192b85262344dce4f4f64314/internal/gatewayapi/contexts.go#L471-L479

so add a `if` logic to deal with `HTTPRoute`